### PR TITLE
90th Percentile Calculator - 2023 Update

### LINKED ISSUE #2847: 90th Percentile Calculator: Update Weather Data to End of 2023
**Describe the task**
Weather data are now a couple years out of date: we need to update the weather data used by 90th Percentile Calculator to the end of 2023: go back 30 years from 2023 for the stations that do have 30 yr data; otherwise report N/A, perhaps? DEVs to decide

**Acceptance Criteria**
- [ ] Weather data are updated to end of 2023 on Datamart
- [ ] 90th percentile calculations reference the new date range.
- [ ] remove quick deploy (QD = non-permanent) and Environment Canada (EC) stations
- [ ] Include all historical stations including active.

**Additional context**
- Documentation for the Datamart and API options for accessing BCWS weather data can be found here: https://bcgov.sharepoint.com/:b:/r/teams/01324/Shared%20Documents/General/_Prevention%20Products%20(90th%20Percentile)/Percentile%20Calculator/BCWS%20Datamart%20and%20API%20%20V2.pdf?csf=1&web=1&e=QQBryt 
- Don't rely on the pre-existing weather_stations.json file as a source of weather station info.

### PR DESCRIPTION
**Pull request by brettedw**
- Updates data to 2023
- Changes slider range to 30 years
- Adds warning if there is missing years for a station and time range selection
- Resolves #2847 
# Test Links:
[Landing Page](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-3250.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: api/scripts/calculate_percentile_offline.py [hunk 1]
```diff
@@ -2,24 +2,30 @@
...
-RECENT_YEAR = 2019  # the most recent year that has the core fire season recorded
-RANGES = [(1970, RECENT_YEAR), (RECENT_YEAR - 19, RECENT_YEAR),
+RECENT_YEAR = 2023  # the most recent year that has the core fire season recorded
+RANGES = [(RECENT_YEAR - 29, RECENT_YEAR), (RECENT_YEAR - 19, RECENT_YEAR),
           (RECENT_YEAR - 9, RECENT_YEAR)]
 PERCENTILE = 0.9
 NUMBER_OF_DECIMAL_POINT = 5  # for FWI values
 
+# this is the path to the parent folder containing years of BCWS historical data in csv format
+BASE = Path(R'/path/to/csv/parent/BCWS_datamart_historical_wx_obs')
```
Comment (conbrad) [inline api/scripts/calculate_percentile_offline.py:16]: Is it possible read the CSV directly from s3?

Also can we folder each datamart dump by year? e.g. `2023/BCWS_datamart_historical_wx_obs` or some way to understand what the latest datadump is next time we run the script.

File: api/scripts/calculate_percentile_offline.py [hunk 2]
```diff
@@ -127,11 +132,11 @@ def remove_data_outside_fire_season(df, season: dict):
     df.drop(df[after_end].index, inplace=True)
 
 
-def calculate_percentile(df, percentile: float) -> dict:
+def calculate_percentile(df: pd.DataFrame, percentile: float) -> dict:
     """ Calculate percentile """
-    ffmc = df[df['ffmc_valid']].ffmc.quantile(percentile)
-    bui = df[df['bui_valid']].bui.quantile(percentile)
-    isi = df[df['isi_valid']].isi.quantile(percentile)
+    ffmc = df[df['FINE_FUEL_MOISTURE_CODE'].notnull()]['FINE_FUEL_MOISTURE_CODE'].quantile(percentile)
```
Comment (conbrad) [inline api/scripts/calculate_percentile_offline.py:137]: Do we now have records with missing indices?

Comment (brettedw) [reply api/scripts/calculate_percentile_offline.py:137]: Could you elaborate on that?

Comment (conbrad) [reply api/scripts/calculate_percentile_offline.py:137]: Got it, thanks!

File: api/scripts/calculate_percentile_offline.py (hunk 1 above)
Comment (conbrad) [reply api/scripts/calculate_percentile_offline.py:16]: We could also document usage of the script and hold off on this in a follow up task.

**Comment by conbrad** [comment]
Is there an easy way to filter out the stations that don't have data in the dropdown list? Currently the station list is populated from the WF1 stations, but for example station 1142 doesn't have any data.

File: api/scripts/station_csv_to_json.py [hunk 3]
```diff
@@ -2,89 +2,113 @@
...
+        if row['STATION_CODE'] == "447":
             ecodivision_name = "SUB-ARCTIC HIGHLANDS"
         else:
-            ecodivision_name = fetch_ecodivision_name(
-                row[lat], row[long], ECODIVISIONS)
+            ecodivision_name = fetch_ecodivision_name(row['LATITUDE'], row['LONGITUDE'], ecodivisions)
 
         if ecodivision_name is not None:
-            core_season = CORE_SEASONS[ecodivision_name]['core_season']
+            core_season = core_seasons.get(ecodivision_name, {}).get('core_season', {})
```
[the remark this answers is not in the document]
Comment (brettedw) [reply api/scripts/station_csv_to_json.py:51]: I should improve this for sure, even though currently I don't think there's a chance of it happening, the logic isn't great here. 
This also lead me to noticing that there is basically 2 defaults in the code as it was (and still currently is) written. In the `ecodivisions_core_seasons.json` there is a default `core_season` ending on Sept. 15, while the Python default is Aug. 31....

**Review by conbrad — APPROVED** [review]
Nice work!

**Review by dgboss — APPROVED** [review]
Looks good!

**Pull request merged**